NVidia: Partial rewrite, Explicit sync, Addition of missing information (#601)
hyprwm/hyprland-wiki: pages/Nvidia/_index.md

named 'nvidia' (or 'nvidia-dkms' to use with custom linux kernels) which is
under active development, the legacy drivers 'nvidia-3xxxx' for older cards
which Nvidia no longer actively supports, and the 'nvidia-open' driver which is
currently an alpha stage attempt to open source a part of their closed source
driver for newer cards.

If your card is supported by the proprietary nvidia drivers, there is a very
high chance that you will most likely want to use those. Benefits include but
are not limited to:

- (Much) better gaming performance
- Better power management on recent GPUs

However, keep in mind that if the proprietary Nvidia drivers do not work
properly on your computer, the Nouveau driver might work fine. This will
most likely be the case for
[older cards](https://wiki.archlinux.org/title/NVIDIA#Unsupported_drivers)

## How to get Hyprland to possibly run on Nvidia (Proprietary)

Install the correct headers package for your current kernel. For the example of
the `linux-zen` kernel on Arch Linux, this package would be `linux-zen-headers`.

Install the required nvidia packages. For most cases, this would be `nvidia-dkms`
(or `nvidia-open-dkms` for the open source ones),

> and `nvidia-utils`, If you'd like to game via Wine or even natively, it would be

Reviewer: ```suggestion
and `nvidia-utils`. If you'd like to game via Wine or even natively, it would be
```

{{< callout >}}

Even if your GPU is listed as supported by the `nvidia-open-dkms` driver, at this
point in time, it is still not up to feature parity with the proprietary drivers.
One issue with the open drivers is that it could cause problems with suspend in
general, let that be closing the lid on your laptop or by manually triggering one.
Overall, you'd be better off with `nvidia-dkms` right now, but Hyprland should work
similarly between the two.

{{< /callout >}}

Next up, you need to enable modeset for nvidia, this can be done via editing
the kernel paramaters for your bootloader.

For people using [systemd-boot](https://wiki.archlinux.org/title/systemd-boot)
you can do this by adding `nvidia_drm.modeset=1` to the end of
`/boot/loader/entries/arch.conf`.

For people using
[GRUB](https://wiki.archlinux.org/title/GRUB) you can do this by adding
`nvidia_drm.modeset=1` to the end of `GRUB_CMDLINE_LINUX_DEFAULT=` in
`/etc/default/grub`, then running `sudo grub-mkconfig -o /boot/grub/grub.cfg`.

For others check out
[kernel parameters](https://wiki.archlinux.org/title/Kernel_parameters) and how
to add `nvidia_drm.modeset=1` to your specific bootloader.

{{< callout >}}

> there has been a lot of confusion on whether or not `nvidia_drm` or `nvidia-drm`
> is correct for this kernel paramter. However, it has been confirmed that either
> of these will work, so do not fret if you already have one or the other in your
> kernel parameters.

Reviewer: ```suggestion
There has been a lot of debate on which of `nvidia_drm` or `nvidia-drm`
is correct for this kernel paramter. It has been confirmed that either
of these will work.
```

Less informal imo.

env = WLR_NO_HARDWARE_CURSORS,1
```

> Last but not least, you'll also need to install a few packages to get some apps
> to function natively with Wayland, you can have a look at [the Master Tutorial](https://wiki.hyprland.org/Getting-Started/Master-Tutorial/#force-apps-to-use-wayland)

Reviewer: ```suggestion
Lastly, you also need to install a few packages to get some apps
to function natively with Wayland. Have a look at [the Master Tutorial](https://wiki.hyprland.org/Getting-Started/Master-Tutorial/#force-apps-to-use-wayland).
```

We want to be concise and to the point.

- Enable the services `nvidia-suspend.service`, `nvidia-hibernate.service` and
`nvidia-resume.service`

- Add `nvidia.NVreg_PreserveVideoMemoryAllocations=1` to your kernel parameters if
you don't have it already. This will solve issues with corrupted desktop / videos
after waking.

- You can check the README to get it working for Firefox.

## Other issues to look out for

### Regarding environment variables

- If you encounter crashes in Firefox, remove the line
`env = GBM_BACKEND,nvidia-drm`.

- If you face problems with Discord windows not displaying or screen sharing not
working in Zoom, remove or comment the line
`env = __GLX_VENDOR_LIBRARY_NAME,nvidia`.

### How to possibly get multi-monitor working with hybrid graphics

On a hybrid graphics device (a laptop with
both an Intel and an Nvidia GPU), you will need to remove the `optimus-manager`
package if installed (disabling the service does not work). You also need to
change your BIOS settings from hybrid graphics to discrete graphics.

### Fixing flickering in electron apps

This flickering is likely caused by these apps running in XWayland.
To fix the flickering, try running the apps with native Wayland instead.

> For discord, you can try the [vesktop app](https://github.com/Vencord/Vesktop).
> It has a multitude of install options for many distros in their README.
> 
> After installing, you can then enable the wayland backend with this command:
> `echo "--enable-features=UseOzonePlatform --ozone-platform-hint=auto" > ~/.config/vesktop-flags.conf`
> 
> This will add the parameters required to run vesktop with its wayland backend.
> 
> On earlier nvidia driver versions, including 535, you may have to also include
> the `--disable-gpu` and `--disable-gpu-sandbox` flags, but as the names suggest,
> you will lose hardware acceleration for vesktop.
> 
> In other cases like vscodium, you can add the same flags to their respective
> flags files. In vscodium's case, this would be `~/.config/codium-flags.conf`

Reviewer: ```suggestion
For Discord, you can try the [Vesktop app](https://github.com/Vencord/Vesktop).
It has a multitude of install options for many distros in their README.

After installing, you can then enable the Wayland backend with this command:
`echo "--enable-features=UseOzonePlatform --ozone-platform-hint=auto" > ~/.config/vesktop-flags.conf`

This will add the parameters required to run Vesktop with its Wayland backend.

On earlier Nvidia driver versions, including 535, you may have to also include
the `--disable-gpu` and `--disable-gpu-sandbox` flags, but, as the names suggests,
you will lose hardware acceleration for Vesktop.

In other cases like VSCodium, you can add the same flags to their respective
flags files. In VSCodium's case, this would be `~/.config/codium-flags.conf`
```
